src/System.Private.TypeLoader/src/Internal/Runtime/TypeLoader/TypeLoaderEnvironment.ValueTypeSupport.cs

```diff
@@ -0,0 +1,181 @@
+
+            if (!RuntimeAugments.IsDynamicType(typeHandle))
+            {
+                return StaticValueTypeGetHashCode(typeHandle, thisObj);
+            }
+            else
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        private int StaticValueTypeGetHashCode(RuntimeTypeHandle typeHandle, ValueType thisObj)
+        {
+            // TODO: we should also use hashcode of the type
+
+            NativeFormatModuleInfo module = ModuleList.GetModuleInfoByHandle(RuntimeAugments.GetModuleFromTypeHandle(typeHandle));
+
+            NativeHashtable layoutHashTable;
+            ExternalReferencesTable externalReferencesLookup;
+
+            if (!GetHashtableFromBlob(module, ReflectionMapBlob.InstanceFieldLayoutHashtable, out layoutHashTable, out externalReferencesLookup))
+                Environment.FailFast("Instance field layout not generated");
+
+            ref byte thisRawData = ref RuntimeAugments.GetRawData(thisObj);
+
+            var enumerator = layoutHashTable.Lookup(typeHandle.GetHashCode());
+
+            //
+            // First try to find instance field layout information in the hashtable.
+            //
+            NativeParser entryParser;
+            while (!(entryParser = enumerator.GetNext()).IsNull)
+            {
+                RuntimeTypeHandle tentativeType = externalReferencesLookup.GetRuntimeTypeHandleFromIndex(entryParser.GetUnsigned());
+                if (!typeHandle.Equals(tentativeType))
+                    continue;
+
+                // Found the entry - find the first non-null field
+                int fieldCount = (int)entryParser.GetUnsigned();
+                for (int i = 0; i < fieldCount; i++)
+                {
+                    RuntimeTypeHandle fieldType = externalReferencesLookup.GetRuntimeTypeHandleFromIndex(entryParser.GetUnsigned());
+                    int fieldOffset = (int)entryParser.GetUnsigned();
+                    
+                    // TODO: the CLR doesn't seem to be boxing and calling GetHashCode on valuetype fields, even if they
+                    //       override GetHashCode? We might be able to get rid of the box here.
+                    object fieldValue = RuntimeAugments.RhBoxAny(ref Unsafe.Add(ref thisRawData, fieldOffset), fieldType);
+                    if (fieldValue != null)
+                        return fieldValue.GetHashCode();
+                }
+
+                // Found no non-null fields
+                return 0;
+            }
+
+            //
+            // If the field layout information isn't present, we can hash the memory
+            //
+            return FastValueTypeGetHashCode(ref thisRawData, typeHandle);
+        }
+
+        private static int FastValueTypeGetHashCode(ref byte location, RuntimeTypeHandle type)
+        {
+            int size = RuntimeAugments.GetValueTypeSize(type);
+            int hashCode = 0;
+
+            for (int i = 0; i < size / 4; i++)
```

Comment:
Same comment -- just the first field.

Reply:
> Same comment -- just the first field.

That's not how CoreCLR does it either:

https://github.com/dotnet/coreclr/blob/4e6435be1f5a77abb875f946fa7bd5dbeb5c3864/src/vm/comutilnative.cpp#L2070-L2089

Reply:
Is size guaranteed to be a multiple of 4? If not, we're including random memory here.

Reply:
> If not, we're including random memory here.

Are you sure? `i` goes from `0` to `size / 4` - we discard the remaining bytes. Again, this is not something I invented, I just stole it from CoreCLR.

Reply:
Whoops, you're right about that. No danger of reading extra memory, though I wonder if it means all 1-3 byte structs end up with 0. This hash code is really really poor. I wonder if we should consider adopting the new HashCode type at some point.